teensy4/usb_audio.cpp

```diff
@@ -323,114 +323,168 @@ static void copy_from_buffers(uint32_t *dst, int16_t *left, int16_t *right, unsi
 	}
 }
 
+/*
+ * On update(), we just receive the audio blocks and keep a set of pointers
+ * to them. The USB transmit callback will then copy them to the transmit buffer
+ * and release them at some point in the future.
+ */
 void AudioOutputUSB::update(void)
 {
-	audio_block_t *left, *right;
-
-	// TODO: we shouldn't be writing to these......
-	//left = receiveReadOnly(0); // input 0 = left channel
-	//right = receiveReadOnly(1); // input 1 = right channel
-	left = receiveWritable(0); // input 0 = left channel
-	right = receiveWritable(1); // input 1 = right channel
-	if (usb_audio_transmit_setting == 0) {
-		if (left) release(left);
-		if (right) release(right);
-		if (left_1st) { release(left_1st); left_1st = NULL; }
-		if (left_2nd) { release(left_2nd); left_2nd = NULL; }
-		if (right_1st) { release(right_1st); right_1st = NULL; }
-		if (right_2nd) { release(right_2nd); right_2nd = NULL; }
+	audio_block_t* chans[AUDIO_CHANNELS];
+	int i;
+	
+
+	// get the audio data
+	for (i=0;i<AUDIO_CHANNELS;i++)
+		chans[i] = receiveReadOnly(i);
+	
+	if (usb_audio_transmit_setting == 0) // not transmitting: dump all audio data
+	{
+		for (i=0;i<AUDIO_CHANNELS;i++)
+		{
+			if (NULL != chans[i]) 
+				release(chans[i]);
+			if (NULL != outgoing[i]) 
+			{
+				release(outgoing[i]);
+				outgoing[i] = NULL;
+			}
+				
+			if (NULL != ready[i]) 
+			{
+				release(ready[i]);
+				ready[i] = NULL;
+			}
+		}
 		offset_1st = 0;
-		return;
 	}
-	if (left == NULL) {
-		left = allocate();
-		if (left == NULL) {
-			if (right) release(right);
-			return;
+	else
+	{
+		// ensure every channel has a real audio block, even if it's silent
+		for (i=0;i<AUDIO_CHANNELS;i++)
+		{
+			if (NULL == chans[i]) // sent NULL: make implied silence into real data
+			{
+				chans[i] = allocate();
+				if (NULL != chans[i])
+					memset(chans[i]->data, 0, sizeof(chans[i]->data));
+				else
+					break; // no block available, exit early
```

Comment:
```suggestion
				if (NULL != chans[i]) {
					memset(chans[i]->data, 0, sizeof(chans[i]->data));
				} else {
					for (int j=0; i<AUDIO_CHANNELS; j++) {
                        if (NULL != chans[j]) release(chans[j]);
                    };
                    return; // no block available, exit early
                }
```

Personally, something like this feels more obvious to me (getting rid of that `if (i >= AUDIO_CHANNELS) { ... } else { ... }` block), just as far as flow legibility is concerned.

Reply:
I have a serious dislike of the "early return" style of error handling! I find it tends to make the code hard to maintain or modify, because you have to catch _all_ the instances and figure out if anything needs unwinding before the return, e.g. a disabled interrupt. It also make it hard to test: I put pin toggles at the start and end of usb_audio_receive_callback(), and they didn't work as expected because of all the early returns - in order to instrument it on my 'scope I ended up wrapping it in another function...

Using the final value of i is a bit obscure, I'll admit; I should have created an "all OK" flag and used that, really.

Reply:
I would absolutely tend to agree with you, coming from mainly doing Rust myself. Unfortunately the rest of this codebase doesn't, and maybe the answer is to simply have that philosophy for this USB audio class. Be the change you want to see in the world ;). Please ignore this suggestion, and I can make the rest of the code consistent with this rule.

Reply:
Too true ... and I don't feel the need to change it unless I'm doing heavy revisions, I just sigh and move on...